File: src/main/java/io/mfedirko/learning/LearningRepository.java
package io.mfedirko.learning;

import org.springframework.cache.annotation.Cacheable;

import java.time.LocalDate;
import java.time.Year;
import java.util.List;

public interface LearningRepository {

    List<Lesson> findLessons(Year year);

    Lesson getLesson(long creationTimeMillis);

    long createLesson(CreateLessonForm lesson);

    void updateLesson(UpdateLessonForm lesson, long creationTimeMillis);

    void deleteLesson(long creationTimeMillis);
}

File: src/main/kotlin/io/mfedirko/learning/LearningRepository.kt
package io.mfedirko.learning

import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.Cacheable
import java.time.Year

interface LearningRepository {
    @Cacheable
    fun findLessons(year: Year): List<Lesson>
    fun getLesson(id: Any): Lesson?
    @CacheEvict(allEntries = true)
    fun createLesson(lesson: CreateLessonForm): Any
    @CacheEvict(allEntries = true)
    fun updateLesson(lesson: UpdateLessonForm, id: Any)
    @CacheEvict(allEntries = true)
    fun deleteLesson(id: Any)
}